pravega/pravega-operator
# Add certs overrides and improve chart resource metadata

### Change log description

This PR adds overrides to the Pravega Operator webhook certs as well as the ability to use self-signed or not. Metadata throughout the chart is also enhanced with explicit namespaces.

### Purpose of the change

These changes are needed to support non-self-signed custom certificates as well as generating the resources using `helm template`.

### What the code does

Listed below are all the changes done in the Pravega Operator charts:
- Added explicit `{{ .Release.Namespace }}` namespaces in all namespace-bound resources (to support helm template)
- Moved all certificate related resources into the same file (for readability)
- Added BookkeeperCluster permissions to the operator roles (to support upgrades)
- Added standard best-practice Helm labels (for ease of CLI commands)
- Modified the `fullname` template to allow for full override
- Fully support non-default namespaces by interpolating the release namespace where it was using `default`
- Add support for external non-self-signed certificates

### How to verify it

We can use `helm template` to verify that all the changes are occurring as they are supposed to. We can see among other things the `metadata.namespace` being templated in, as well as the best-practice Helm labels correctly generating.

#### With self-signed certs:
`values.yaml`:
```
webhookCert:
  selfsigned: true
```

Notice that no custom Secret is taken and that the Issuer is to create its own self-signed certificate. We can also notice the Certificate using the correct namespaces instead of `default`.
```
# Source: pravega-operator/templates/certificate.yaml
apiVersion: certmanager.k8s.io/v1alpha1
kind: Issuer
metadata:
  name: selfsigned-cert
  namespace: pravega
  labels:
    app.kubernetes.io/name: pravega-operator
    app.kubernetes.io/version: "0.5.0"
    helm.sh/chart: "pravega-operator-0.5.0"
spec:
  selfSigned: {}
---
# Source: pravega-operator/templates/certificate.yaml
apiVersion: certmanager.k8s.io/v1alpha1
kind: Certificate
metadata:
  name: selfsigned-cert
  namespace: pravega
  labels:
    app.kubernetes.io/name: pravega-operator
    app.kubernetes.io/version: "0.5.0"
    helm.sh/chart: "pravega-operator-0.5.0"
spec:
  isCA: true
  duration: 8760h
  renewBefore: 360h
  secretName: selfsigned-cert-tls
  commonName: pravega-webhook-svc.pravega.svc.cluster.local
  dnsNames:
    - pravega-webhook-svc
    - pravega-webhook-svc.pravega.svc.cluster.local
    - pravega-webhook-svc.pravega.svc
  issuerRef:
    name: selfsigned-cert
```

#### With custom certs:
`values.yaml`:
```
webhookCert:
  selfsigned: false
  crt: fake_cert_crt_LS0tLS1CRUdJTiBDRVJUSUZJQ0FURS0tLS0tCk1JSURyRENDQX
  key: fake_cert_key_LS0tLS1CRUdJTiBDRVJUSUZJQ0FURS0tLS0tCk1JSURyRENDQX
```

Notice that a Secret for the custom certificate Secret gets created and that the Issuer takes that Secret instead of generating its own. We can also notice the Certificate using the correct namespaces instead of `default`.
```
# Source: pravega-operator/templates/certificate.yaml
apiVersion: v1
kind: Secret
metadata:
  name: pravega-operator-key-pair
  namespace: pravega
type: kubernetes.io/tls
data:
  tls.crt: fake_cert_crt_LS0tLS1CRUdJTiBDRVJUSUZJQ0FURS0tLS0tCk1JSURyRENDQX
  tls.key: fake_cert_key_LS0tLS1CRUdJTiBDRVJUSUZJQ0FURS0tLS0tCk1JSURyRENDQX
---
# Source: pravega-operator/templates/certificate.yaml
apiVersion: certmanager.k8s.io/v1alpha1
kind: Issuer
metadata:
  name: selfsigned-cert
  namespace: pravega
  labels:
    app.kubernetes.io/name: pravega-operator
    app.kubernetes.io/version: "0.5.0"
    helm.sh/chart: "pravega-operator-0.5.0"
spec:
  ca:
    secretName: pravega-operator-key-pair
---
# Source: pravega-operator/templates/certificate.yaml
apiVersion: certmanager.k8s.io/v1alpha1
kind: Certificate
metadata:
  name: selfsigned-cert
  namespace: pravega
  labels:
    app.kubernetes.io/name: pravega-operator
    app.kubernetes.io/version: "0.5.0"
    helm.sh/chart: "pravega-operator-0.5.0"
spec:
  isCA: true
  duration: 8760h
  renewBefore: 360h
  secretName: selfsigned-cert-tls
  commonName: pravega-webhook-svc.pravega.svc.cluster.local
  dnsNames:
    - pravega-webhook-svc
    - pravega-webhook-svc.pravega.svc.cluster.local
    - pravega-webhook-svc.pravega.svc
  issuerRef:
    name: selfsigned-cert
```

#### Testing with PravegaCluster

Additionally, we can deploy a full PravegaCluster to verify that it deploys without any errors:
```
pravega   pravega-controller-74865546f-z6csr         1/1     Running     0          62m
pravega   pravega-grafana-0                          2/2     Running     0          62m
pravega   pravega-influxdb-0                         1/1     Running     0          62m
pravega   pravega-segment-store-0                    1/1     Running     0          62m
pravega   pravega-operator-6557bf68c6-whw5g          1/1     Running     0          67m
```
